lib/backends/webgl/op-resolve-rules.ts

```diff
@@ -0,0 +1,100 @@
+import * as reduceOps from './ops/reduce';
+import {WebGLReshape} from './ops/reshape';
+import {WebGLSlice} from './ops/slice';
+import {WebGLSoftmax} from './ops/softmax';
+import {WebGLSplit} from './ops/split';
+import {WebGLSqueeze} from './ops/squeeze';
+import {WebGLSum} from './ops/sum';
+import {WebGLTile} from './ops/tile';
+import {WebGLTranspose} from './ops/transpose';
+import * as unaryOps from './ops/unary-op';
+import {WebGLUnsqueeze} from './ops/unsqueeze';
+
+export const WEBGL_OP_RESOLVE_RULES: ReadonlyArray<OpSet.ResolveRule> = [
+  ['Abs', '', '6+', () => new unaryOps.WebGLUnaryOp(NUMBER_TYPES, unaryOps.glslAbs())],
+  ['Acos', '', '7+', () => new unaryOps.WebGLUnaryOp(FLOAT_TYPES, unaryOps.glslAcos())],
+  ['Add', '', '7+', () => new binaryOps.WebGLBinaryOp(NUMBER_TYPES, binaryOps.glslAdd())],
+  ['And', '', '7+', () => new binaryOps.WebGLBinaryOp(['bool'], binaryOps.glslAnd())],
+  ['Asin', '', '7+', () => new unaryOps.WebGLUnaryOp(FLOAT_TYPES, unaryOps.glslAsin())],
+  ['Atan', '', '7+', () => new unaryOps.WebGLUnaryOp(FLOAT_TYPES, unaryOps.glslAtan())],
+  ['AveragePool', '', '7+', () => new WebGLAveragePool()],
+  ['BatchNormalization', '', '7+', () => new WebGLBatchNormalization()],
+  ['Ceil', '', '6+', () => new unaryOps.WebGLUnaryOp(FLOAT_TYPES, unaryOps.glslCeil())],
+  ['Concat', '', '4+', () => new WebGLConcat()],
+  ['Conv', '', '1+', () => new WebGLConv()],
+  ['Cos', '', '7+', () => new unaryOps.WebGLUnaryOp(FLOAT_TYPES, unaryOps.glslCos())],
+  ['Div', '', '7+', () => new binaryOps.WebGLBinaryOp(NUMBER_TYPES, binaryOps.glslDiv())],
+  ['Dropout', '', '7+', () => new WebGLDropout()],
+  ['Equal', '', '7+', () => new binaryOps.WebGLBinaryOp(NUMBER_TYPES, binaryOps.glslEqual(), undefined, 'bool')],
+  ['Elu', '', '6+', () => new WebGLElu()],
+  ['Exp', '', '6+', () => new unaryOps.WebGLUnaryOp(FLOAT_TYPES, unaryOps.glslExp())],
+  ['Flatten', '', '1+', () => new WebGLFlatten()],
+  ['Floor', '', '6+', () => new unaryOps.WebGLUnaryOp(FLOAT_TYPES, unaryOps.glslFloor())],
+  ['Gather', '', '1+', () => new WebGLGather()],
+  ['Gemm', '', '7+', () => new WebGLGemm()],
+  ['GlobalAveragePool', '', '1+', () => new WebGLGlobalAveragePool()],
+  ['GlobalMaxPool', '', '1+', () => new WebGLGlobalMaxPool()],
+  ['Greater', '', '7+', () => new binaryOps.WebGLBinaryOp(NUMBER_TYPES, binaryOps.glslGreater(), undefined, 'bool')],
+  ['Identity', '', '1+', () => new unaryOps.WebGLUnaryOp(NUMBER_TYPES, unaryOps.glslIdentity())],
+  ['ImageScaler', '', '1+', () => new WebGLImageScaler()],
+  ['LeakyRelu', '', '6+', () => new WebGLLeakyRelu()],
+  ['Less', '', '7+', () => new binaryOps.WebGLBinaryOp(NUMBER_TYPES, binaryOps.glslLess(), undefined, 'bool')],
+  ['Log', '', '6+', () => new unaryOps.WebGLUnaryOp(FLOAT_TYPES, unaryOps.glslLog())],
+  ['MatMul', '', '1+', () => new WebGLMatMul()],
+  ['MaxPool', '', '1-7', () => new WebGLMaxPool()],  // TODO: support new attributes for MaxPool-8 and MaxPool-10
```

Comment:
Developer-centric question - I am guessing the way we support different versions of an op (where there is majority overlap across versions) is we just extend the current working class and override appropriate methods so as to align with the new spec ?

Reply:
there are 2 ways.
one is to extend the existing implementation, _if the new version is compatible with the old one according to the spec_. (eg. add a new optional attribute, whose default value is the same of the old behavior)
one is to have different implementation, _if the new version has a different behavior, or it is necessary to do so due to performance concern_.